server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -348,61 +349,71 @@ protected void doInternalExecute(Task task, BulkRequest bulkRequest, Executor ex
-        Set<String> dataStreamsToBeRolledOver = featureService.clusterHasFeature(state, LazyRolloverAction.DATA_STREAM_LAZY_ROLLOVER)
-            ? indices.keySet().stream().filter(target -> {
-                DataStream dataStream = state.metadata().dataStreams().get(target);
-                return dataStream != null && dataStream.rolloverOnWrite();
-            }).collect(Collectors.toSet())
-            : Set.of();
-
-        // Step 4: create all the indices that are missing, if there are any missing. start the bulk after all the creates come back.
         createMissingIndicesAndIndexData(
             task,
             bulkRequest,
             executor,
             listener,
             indicesToAutoCreate,
             dataStreamsToBeRolledOver,
-            indicesThatCannotBeCreated,
+            failureStoresToBeRolledOver,
             startTime
         );
     }
 
-    /*
-     * This method is responsible for creating any missing indices, rolling over a data stream when needed and then
-     *  indexing the data in the BulkRequest
+    private void populateMissingTargets(
+        BulkRequest bulkRequest,
+        Map<String, Boolean> indicesToAutoCreate,
+        Set<String> dataStreamsToBeRolledOver,
+        Set<String> failureStoresToBeRolledOver
+    ) {
+        Map<String, Boolean> indexExistence = new HashMap<>();
+        ClusterState state = clusterService.state();
+        boolean lazyRolloverFeature = featureService.clusterHasFeature(state, LazyRolloverAction.DATA_STREAM_LAZY_ROLLOVER);
+        boolean lazyRolloverFailureStoreFeature = featureService.clusterHasFeature(state, LazyRolloverAction.FAILURE_STORE_LAZY_ROLLOVER);
+
+        for (DocWriteRequest<?> request : bulkRequest.requests) {
+            if (request.opType() == OpType.DELETE
+                && request.versionType() != VersionType.EXTERNAL
+                && request.versionType() != VersionType.EXTERNAL_GTE) {
+                continue;
+            }
+            Boolean indexExists = indexExistence.get(request.index());
+            if (indexExists == null) {
+                indexExists = indexNameExpressionResolver.hasIndexAbstraction(request.index(), state);
+                indexExistence.put(request.index(), indexExists);
+            }
+            if (indexExists == false && request.isRequireAlias() == false) {
+                Boolean requiresDataStream = indicesToAutoCreate.get(request.index());
+                if (requiresDataStream == null || (requiresDataStream == false && request.isRequireDataStream())) {
+                    indicesToAutoCreate.put(request.index(), request.isRequireDataStream());
+                }
+            }
+            if (lazyRolloverFeature) {
+                DataStream dataStream = state.metadata().dataStreams().get(request.index());
+                if (dataStream != null) {
+                    // TODO: Do we only want to roll over data streams on _index_ requests?
```

Comment:
Writes to the failure store are only ever index operations. Deletes are not considered, and updates should be disallowed on data streams. I think it's safe to make this assumption here for writing to the failure store.

Reply:
> Deletes are not considered

In my current logic (`writeToFailureStore == false && dataStream.getBackingIndices().isRolloverOnWrite()`), a delete request to a data stream that is marked for lazy rollover (i.e. that backing indices are marked for rollover, not the failure store) will roll over the data stream. My TODO question was directed toward that specific scenario.

Reply:
That scenario is not possible.

Using the data stream name you cannot delete a document: `only write ops with an op_type of create are allowed in data streams`. 

Using the backing index name you can delete the document but that will not trigger a rollover because the `rolloverOnWrite` is on data stream level.

Does this answer your TODO?

Reply:
Yep definitely, thanks both!